src/engine/BinaryExport.cpp

```diff
@@ -0,0 +1,387 @@
+// Copyright 2025, University of Freiburg
+// Chair of Algorithms and Data Structures
+// Authors: Johannes Kalmbach <kalmbach@cs.uni-freiburg.de>
+//          Robin Textor-Falconi <textorr@cs.uni-freiburg.de>
+
+#include "engine/BinaryExport.h"
+
+#include <absl/functional/bind_front.h>
+
+#include "ExportQueryExecutionTrees.h"
+// TODO<joka921> Can we get further type erasure here to not include this
+// detail?
+#include "util/Serializer/ByteBufferSerializer.h"
+#include "util/Serializer/SerializeOptional.h"
+#include "util/Serializer/SerializeString.h"
+#include "util/Serializer/SerializeVector.h"
+#include "util/http/HttpClient.h"
+
+using CancellationHandle = ad_utility::SharedCancellationHandle;
+namespace ad_utility::serialization {
+// TODO<joka921> Move this to a global file.
+template <typename F>
+class ReadViaCallableSerializer {
+ public:
+  using SerializerType = ReadSerializerTag;
+
+ private:
+  F readFunction_;
+
+ public:
+  explicit ReadViaCallableSerializer(F readFunction)
+      : readFunction_{std::move(readFunction)} {};
+  void serializeBytes(char* bytePointer, size_t numBytes) {
+    readFunction_(bytePointer, numBytes);
+  }
+
+  ReadViaCallableSerializer(const ReadViaCallableSerializer&) noexcept = delete;
+  ReadViaCallableSerializer& operator=(const ReadViaCallableSerializer&) =
+      delete;
+  ReadViaCallableSerializer(ReadViaCallableSerializer&&) noexcept = default;
+  ReadViaCallableSerializer& operator=(ReadViaCallableSerializer&&) noexcept =
+      default;
+};
+}  // namespace ad_utility::serialization
+namespace {
+// Return a `std::string_view` wrapping the passed value.
+std::string_view raw(const std::integral auto& value) {
+  return std::string_view{reinterpret_cast<const char*>(&value), sizeof(value)};
+}
+}  // namespace
+
+namespace qlever::binary_export {
+
+// _____________________________________________________________________________
+bool BinaryExportHelpers::isTrivial(Id id) {
+  auto datatype = id.getDatatype();
+  return datatype == Datatype::Undefined || datatype == Datatype::Bool ||
+         datatype == Datatype::Int || datatype == Datatype::Double ||
+         datatype == Datatype::Date || datatype == Datatype::GeoPoint ||
+         datatype == Datatype::EncodedVal;
+}
+
+// _____________________________________________________________________________
+std::string StringMapping::flush(const Index& index) {
+  numProcessedRows_ = 0;
+  std::vector<std::string> sortedStrings;
+  sortedStrings.resize(stringMapping_.size());
+  for (auto& [oldId, newId] : stringMapping_) {
+    auto type = oldId.getDatatype();
+    if (type == Datatype::LocalVocabIndex) {
+      sortedStrings[newId] =
+          oldId.getLocalVocabIndex()->toStringRepresentation();
+    } else {
+      // TODO<joka921, RobinTF>, make the list exhaustive.
```

Comment:
Exhaustive means that we need to also handle `TextRecordIndex`, `WordVocabIndex` and `BlankNodeIndex`.